tensorflow-ops/src/TensorFlow/Gradient.hs

```diff
@@ -108,6 +109,15 @@ type GradientCompatible a =
     -- TODO(fmayle): MaxPoolGrad doesn't support Double for some reason.
     (Num a, OneOf '[ Float, Complex Float, Complex Double ] a)
 
+-- | Types that gradients can be taken w.r.t. (namely 'Tensor' and 'Variable').
+class GradientTarget t where
```

Comment:
`targetOutput` is really just `renderedOutput` (except lifted inside `ResourceHandle` wrapper for the non-`Tensor` cases), so I'm thinking of adding a class for `renderedOutput`. One hesitation is that it overlaps with `Rendered` which would likely be confusing to users, but I think we can replace `Rendered` completely with the `renderedOutput` class.

For `targetZeros`, I like your `tf.convert_to_tensor` suggestion after thinking about it more. I'll try that and do `zerosLike (convertToTensor x)`.